homeassistant/components/plugwise/__init__.py

```diff
@@ -1 +1,169 @@
+from homeassistant.helpers.event import async_track_time_interval
+
+from .const import DOMAIN
+
+CONFIG_SCHEMA = vol.Schema({DOMAIN: vol.Schema({})}, extra=vol.ALLOW_EXTRA)
+
+_LOGGER = logging.getLogger(__name__)
+
+SENSOR_PLATFORMS = ["sensor"]
+ALL_PLATFORMS = ["binary_sensor", "climate", "sensor", "switch"]
+
+
+async def async_setup(hass: HomeAssistant, config: dict):
+    """Set up the Plugwise platform."""
+    return True
+
+
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Set up Plugwise Smiles from a config entry."""
+    websession = async_get_clientsession(hass, verify_ssl=False)
+    api = Smile(
+        host=entry.data["host"], password=entry.data["password"], websession=websession
+    )
+
+    try:
+        connected = await api.connect()
+
+        if not connected:
+            _LOGGER.error("Unable to connect to Smile")
+            raise PlatformNotReady
+
+    except Smile.PlugwiseError:
+        _LOGGER.error("Error while communicating to device")
+        raise PlatformNotReady
+
+    except asyncio.TimeoutError:
+        _LOGGER.error("Timeout while connecting to Smile")
+        raise PlatformNotReady
+
+    if api.smile_type == "power":
+        update_interval = timedelta(seconds=10)
+    else:
+        update_interval = timedelta(seconds=60)
+
+    api.get_all_devices()
+
+    hass.data.setdefault(DOMAIN, {})[entry.entry_id] = {
+        "api": api,
+        "updater": SmileDataUpdater(
+            hass, "device", entry.entry_id, api, "full_update_device", update_interval
+        ),
+    }
+
+    device_registry = await dr.async_get_registry(hass)
+    device_registry.async_get_or_create(
+        config_entry_id=entry.entry_id,
+        identifiers={(DOMAIN, api.gateway_id)},
+        manufacturer="Plugwise",
+        name=entry.title,
+        model=f"Smile {api.smile_name}",
+        sw_version=api.smile_version[0],
+    )
+
+    single_master_thermostat = api.single_master_thermostat()
+    platforms = ALL_PLATFORMS
+    if single_master_thermostat is None:
+        platforms = SENSOR_PLATFORMS
+
+    for component in platforms:
+        hass.async_create_task(
+            hass.config_entries.async_forward_entry_setup(entry, component)
+        )
+
+    async def async_refresh_all(_):
+        """Refresh all Smile data."""
+        for info in hass.data[DOMAIN].values():
+            await info["updater"].async_refresh_all()
+
+    hass.services.async_register(DOMAIN, "update", async_refresh_all)
+
+    return True
+
+
+async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry):
+    """Unload a config entry."""
+    unload_ok = all(
+        await asyncio.gather(
+            *[
+                hass.config_entries.async_forward_entry_unload(entry, component)
+                for component in ALL_PLATFORMS
+            ]
+        )
+    )
+    if unload_ok:
+        hass.data[DOMAIN].pop(entry.entry_id)
+
+    return unload_ok
+
+
+class SmileDataUpdater:
```

Comment:
It looks like this can be replaced with `DataUpdateCoordinator`

https://developers.home-assistant.io/docs/integration_fetching_data/#polling-api-endpoints

Reply:
Working on this, tnx!

Reply:
Commit 09a2a1c is the rewrite to DataUpdateCoordinator, we are preparing the other platforms the same way now while this PR is still in review. Tnx for the trigger @bdraco

Reply:
Marking this one as resolved as per the below thread.